src/ngraph/op/tensor_iterator.cpp

```diff
@@ -0,0 +1,303 @@
+                                                                     int64_t end,
+                                                                     int64_t axis)
+    : OutputDescription(body_value, output_index)
+    , m_start(start)
+    , m_stride(stride)
+    , m_part_size(part_size)
+    , m_end(end)
+    , m_axis(axis)
+{
+}
+
+shared_ptr<op::TensorIterator::OutputDescription>
+    op::TensorIterator::ConcatOutputDescription::copy() const
+{
+    return make_shared<ConcatOutputDescription>(
+        m_body_value, m_output_index, m_start, m_stride, m_part_size, m_end, m_axis);
+}
+
+op::TensorIterator::BodyOutputDescription::BodyOutputDescription(const Output<Node>& body_value,
+                                                                 uint64_t output_index,
+                                                                 int64_t iteration)
+    : OutputDescription(body_value, output_index)
+    , m_iteration(iteration)
+{
+}
+
+shared_ptr<op::TensorIterator::OutputDescription>
+    op::TensorIterator::BodyOutputDescription::copy() const
+{
+    return make_shared<BodyOutputDescription>(m_body_value, m_output_index, m_iteration);
+}
+
+Input<Node> op::TensorIterator::input_for_value(const Output<Node>& value)
+{
+    for (auto input : inputs())
+    {
+        if (input.get_source_output() == value)
+        {
+            return input;
+        }
+    }
+    auto input_index = get_input_size();
+    set_argument(input_index, value);
+    return Input<Node>(this, input_index);
+}
+
+void op::TensorIterator::set_sliced_input(const std::shared_ptr<op::Parameter>& body_parameter,
+                                          const Output<Node>& value,
+                                          int64_t start,
+                                          int64_t stride,
+                                          int64_t part_size,
+                                          int64_t end,
+                                          int64_t axis)
+{
+    m_input_descriptions.push_back(make_shared<SliceInputDescription>(
+        input_for_value(value).get_index(), body_parameter, start, stride, part_size, end, axis));
+}
+
+void op::TensorIterator::set_initialized_input(const std::shared_ptr<Parameter>& body_parameter,
+                                               const Output<Node>& initial_value,
+                                               const Output<Node>& successive_value)
+{
+    m_input_descriptions.push_back(make_shared<BodyConnectionInputDescription>(
+        input_for_value(initial_value).get_index(), body_parameter, successive_value));
+}
+
+Output<Node> op::TensorIterator::get_iter_value(const Output<Node>& body_value, int64_t iteration)
+{
+    auto output_index = get_output_size();
+    m_output_descriptions.push_back(
+        make_shared<BodyOutputDescription>(body_value, output_index, iteration));
+    set_output_size(output_index + 1);
+    return Output<Node>(shared_from_this(), output_index);
```

Comment:
Can we return the data as `op :: Result` instead of` `Output <Node>`? TensorIterator will look like a full ngraph function with input in the form of `Parameter` and output in the form of` op :: Result`.

Reply:
It's not a function right now.  `Result` is more like a stable part of `Function` to hang a piece of graph on so that splicing is simple that it is an op.

Reply:
@diyessi, we would like to have Result op for two purposes: (1) as a clear indication of output in the IR v10 and (2) as an explicit point to connect back-edge and preserve intermediate results in a body to avoid any fusion that can remove this intermediate tensor. So we need them as explicit Result ops. And why do you think that this is not yet a function -- it is really a function, it is separated from an exterior function, it has parameters and results, but currently only Parameters are explicitly materialized, our proposal is to do for Results as well.

Reply:
@ayzhuang @diyessi TensorIterator is critical to us. Can we resolve this question this week?